### PR TITLE
[ndslice] allow init elem for slice()

### PR DESCRIPTION
**Pull request by wilzbach**
ping @9il - I added a changelog entry to show the fancy new allocation support in ndslice.
It got a bit bigger, so I put the unittest in `package.d` to show them - I think they can be quite useful to starters too?

Moreover two small fixes:

- `T` can be inferred for makeSlice when used with an init element (we do this for `slice` too)
- `slice` didn't allow an `init` - this is confusing for the user (see [discussion at mir](https://github.com/libmir/mir/issues/202))

### CODE REVIEW

**Comment by 9il** [comment]
I am working on that XDD

**Comment by 9il** [comment]
Don't close it please. Will use your text )

**Comment by wilzbach** [comment]
rebased to only contain the addition to `slice`, s.t. it accepts init elements

File: std/experimental/ndslice/slice.d [hunk 1]
```diff
@@ -567,6 +567,19 @@ slice(T,
 }
 
 /// ditto
+auto slice(T,
+    Flag!`replaceArrayWithPointer` replaceArrayWithPointer = Yes.replaceArrayWithPointer,
+    size_t N)(auto ref in size_t[N] lengths, auto ref T init)
+{
+    import std.array : uninitializedArray;
+    immutable len = lengthsProduct(lengths);
+    auto arr = uninitializedArray!(T[])(len);
```
Comment (9il) [inline std/experimental/ndslice/slice.d]: `uninitializedArray` should be used  only for types without elaborate constructor

File: std/experimental/ndslice/slice.d [hunk 2]
```diff
@@ -579,14 +592,31 @@ auto slice(
...
+    static assert(is(typeof(sl) == Slice!(3, int*)));
 
     // creates duplicate using `slice`
-    auto dup = .slice(slice);
-    assert(dup == slice);
+    auto dup = sl.slice;
+    assert(dup == sl);
+}
+
+///
```
Comment (9il) [inline std/experimental/ndslice/slice.d]: please split this into 3 unittests

File: std/experimental/ndslice/slice.d [hunk 3]
```diff
@@ -579,14 +592,31 @@ auto slice(
...
+    assert(sl.elementsCount == 2 * 3);
+    assert(sl[1, 1] == 5);
+
+    import std.experimental.ndslice.selection: iotaSlice;
+    auto slIota = iotaSlice(2, 3).slice;
+    assert(slIota == [[0, 1, 2], [3, 4, 5]]);
+
+    // normal ranges can be used too
+    import std.range: iota;
+    auto slRange = iota(6).sliced(2, 3).slice;
```
Comment (9il) [inline std/experimental/ndslice/slice.d]: This is for `sliced` not for `slice`

Comment (9il) [inline std/experimental/ndslice/slice.d]: just remove this example. the previous one already show this feature

File: std/experimental/ndslice/slice.d [hunk 4]
```diff
@@ -579,14 +592,31 @@ auto slice(
 ///
 pure nothrow unittest
 {
-    auto slice = slice!int(5, 6, 7);
-    assert(slice.length == 5);
-    assert(slice.elementsCount == 5 * 6 * 7);
-    static assert(is(typeof(slice) == Slice!(3, int*)));
+    auto sl = slice!int(5, 6, 7);
```
Comment (9il) [inline std/experimental/ndslice/slice.d]: btw, i like `slice` more)

Comment (9il) [inline std/experimental/ndslice/slice.d]: when you read code _slice length_ is better then _sl length_

File: std/experimental/ndslice/slice.d [hunk 5]
```diff
@@ -567,6 +567,19 @@ slice(T,
 }
 
 /// ditto
+auto slice(T,
+    Flag!`replaceArrayWithPointer` replaceArrayWithPointer = Yes.replaceArrayWithPointer,
+    size_t N)(auto ref in size_t[N] lengths, auto ref T init)
```
Comment (9il) [inline std/experimental/ndslice/slice.d]: This looks buggy. Please force user to define a element type

File: std/experimental/ndslice/slice.d (hunk 4 above)
Comment (wilzbach) [inline std/experimental/ndslice/slice.d]: Yep me too, but it's also pretty confusing when one reads the code (is the function or object) now called. how about `v` (for vector)?

Comment (9il) [inline std/experimental/ndslice/slice.d]: `tensor` for ND, N > 2 arrays

File: std/experimental/ndslice/slice.d (hunk 5 above)
Comment (wilzbach) [inline std/experimental/ndslice/slice.d]: you already do the same for `makeSlice` - https://github.com/dlang/phobos/blob/master/std/experimental/ndslice/slice.d#L627

Comment (9il) [inline std/experimental/ndslice/slice.d]: This means that `makeSlice` should be fixed too ;-)

Comment (wilzbach) [inline std/experimental/ndslice/slice.d]: turns out Allocator uses the same template:
https://github.com/dlang/phobos/blob/master/std/experimental/allocator/package.d#L681

```
T[] makeArray(T, Allocator)(auto ref Allocator alloc, size_t length,
    auto ref T init)
```

**Comment by wilzbach** [comment]
@9il did you see these two (existing) templates?

```
T[] makeArray(T, Allocator)(auto ref Allocator alloc, size_t length,
    auto ref T init)
```
(https://github.com/dlang/phobos/blob/master/std/experimental/allocator/package.d#L681)


```
auto makeSlice(T,
    Flag!`replaceArrayWithPointer` replaceArrayWithPointer = Yes.replaceArrayWithPointer,
    Allocator,
    size_t N)(auto ref Allocator alloc, auto ref in size_t[N] lengths, auto ref T init)
```
https://github.com/dlang/phobos/blob/master/std/experimental/ndslice/slice.d#L627

**Comment by 9il** [comment]
ok, I give up)

File: std/experimental/ndslice/slice.d [hunk 6]
```diff
@@ -567,6 +567,26 @@ slice(T,
...
+    {
+        import std.array : uninitializedArray;
+        auto arr = uninitializedArray!(T[])(len);
+    }
+    else
+    {
+        auto arr = new T[len];
+    }
+    auto ret = arr.sliced!replaceArrayWithPointer(lengths);
+    ret[] = init;
```
Comment (9il) [inline std/experimental/ndslice/slice.d]: `arr[] = init;` is faster

**Comment by wilzbach** [comment]
@9il rebased - btw after thinking longer about it, it's not that bad actually. You can still specify the type if you want to by `slice([2, 3], cast(real) 1.0)` etc.

**Comment by 9il** [comment]
Auto-merge toggled on

**Pull request merged**